File: app/src/main/java/com/example/dgtimer/db/CapsuleDao.java
package com.example.dgtimer.db;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface CapsuleDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Capsule capsule);

    @Query("SELECT * FROM capsules ORDER BY major desc, name")
    LiveData<List<Capsule>> getAll();

    @Query("SELECT * FROM capsules WHERE name = :name")
    List<Capsule> getByName(String name);

    @Query("SELECT * FROM capsules WHERE id = :id")
    @Nullable
    Capsule getCapsuleById(int id);

    @Query("SELECT * FROM capsules WHERE name LIKE :name ORDER BY major desc")
    List<Capsule> searchByName(String name);

    @Query("SELECT * FROM capsules WHERE id = :id LIMIT 1")
    LiveData<Capsule> searchById(int id);

    @Query("UPDATE capsules SET major = :major WHERE id = :id")
    void updateMajor(int id, boolean major);

    @Update
    void updateCapsule(Capsule capsule);
}

File: app/src/main/java/com/example/dgtimer/db/CapsuleDao.kt
package com.example.dgtimer.db

import androidx.room.*
import kotlinx.coroutines.flow.Flow

@Dao
interface CapsuleDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCapsule(capsule: Capsule)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCapsules(capsules: List<Capsule>)

    @Query("SELECT * FROM capsules ORDER BY major desc, name")
    fun getAll(): Flow<List<Capsule>?>

    @Query("SELECT * FROM capsules WHERE name = :name")
    fun getByName(name: String): List<Capsule>?

    @Query("SELECT * FROM capsules WHERE id = :id")
    fun getCapsuleById(id: Int): Capsule?

    @Query("SELECT * FROM capsules WHERE name LIKE '%' || :name || '%' ORDER BY major desc")
    fun searchByName(name: String): List<Capsule>?

    @Query("SELECT * FROM capsules WHERE major == 1 ORDER BY name")
    fun getStarredCapsules(): List<Capsule>?

    @Query("SELECT * FROM capsules WHERE id = :id LIMIT 1")
    fun searchById(id: Int): Flow<Capsule?>

    @Query("UPDATE capsules SET major = :major WHERE id = :id")
    fun updateMajor(id: Int, major: Boolean)

    @Update
    fun updateCapsule(capsule: Capsule)
}